mutation_compactor.hh

```diff
@@ -179,8 +179,8 @@ private:
     stop_iteration do_consume(range_tombstone_change&& rtc, Consumer& consumer, GCConsumer& gc_consumer) {
         stop_iteration gc_consumer_stop = stop_iteration::no;
         stop_iteration consumer_stop = stop_iteration::no;
-        if (rtc.tombstone() <= _partition_tombstone) {
-            rtc.set_tombstone({});
+        if (rtc.tombstone() && rtc.tombstone() <= _partition_tombstone) {
+            rtc.set_tombstone(_partition_tombstone);
```

Comment:
Is the bug really here, or in mx sstable writer? That is, I can't think of a reason why a cleared range tombstone change would be illegal under flat_mutation_reader_v2 rules.

This is important, because if the existing code doesn't violate any rules, then we can have the same bad fragment stream produced via a different route and end up with a bad sstable. If that's the case, we need to teach the sstable writer to understand this stream instead.

@tgrabiec ?

Reply:
The relevant bits are [the write path](https://github.com/scylladb/scylla/blob/6a3494442e6e870620b53adf8194fa845d4a5acc/sstables/mx/writer.cc#L1342), where we define the marker written to disk. Without compaction (and clearing the tombstone caused by it) we have three RTCs (rtc1(with tombstone), rtc2(with tombstone), rtc3(empty tombstone)). They go throught the branches (start bound, boundary, end bound). If we clear the second tombstone (as it happens during compaction) we go through branches (start, end, start). The last marker, when read by [the read path](https://github.com/scylladb/scylla/blob/6a3494442e6e870620b53adf8194fa845d4a5acc/sstables/mx/reader.cc#L488), in the uncompacted stream goes through consume_range_tombstone_end path, and in compacted stream goes through consume_range_tombstone_start. That causes the swap from after_all_prefixed to before_all_prefixed.

Reply:
Agreed

Reply:
> The relevant bits are [the write path](https://github.com/scylladb/scylla/blob/6a3494442e6e870620b53adf8194fa845d4a5acc/sstables/mx/writer.cc#L1342), where we define the marker written to disk. Without compaction (and clearing the tombstone caused by it) we have three RTCs (rtc1(with tombstone), rtc2(with tombstone), rtc3(empty tombstone)). They go throught the branches (start bound, boundary, end bound). If we clear the second tombstone (as it happens during compaction) we go through branches (start, end, start). The last marker, when read by [the read path](https://github.com/scylladb/scylla/blob/6a3494442e6e870620b53adf8194fa845d4a5acc/sstables/mx/reader.cc#L488), in the uncompacted stream goes through consume_range_tombstone_end path, and in compacted stream goes through consume_range_tombstone_start. That causes the swap from after_all_prefixed to before_all_prefixed.

What's the answer to the question?

Reply:
I was mostly pointing to the place in code that causes the problem.

The answer to your question depends on whether is it legal to have two subsequent RTCs with empty tombstones.
If it's illegal, then we're producing the wrong stream of mutations to write and the solution proposed in the PR is correct. If it is legal, then the writer/reader pair need to account for that and the stream of mutations after writing to/reading from disk should be identical and the fix needs to go into sstable reader/writer.

Reply:
Is it legal to have two subsequent RTCs with empty tombstones?

Reply:
I have no idea. @bhalevy @tgrabiec @denesb @raphaelsc ?

Reply:
It's redundant, but redundancy is allowed. Only compaction should drop redundancy.

Reply:
We allow other kinds of redundancy, e.g. rtc with the same position

Reply:
I think the bug was exposed by my memtable conversion patch (4f77e74bd4).